File: libs/python/edgecommons/messaging/providers/greengrass/iotcore_subscription_handler.py
import logging
from typing import Callable
from edgecommons.messaging.message import Message
from edgecommons.messaging.providers.greengrass.subscription_handler import (
    SubscriptionHandler,
)

logger = logging.getLogger("IoTCoreSubscriptionHandler")


class IoTCoreSubscriptionHandler(SubscriptionHandler):
    def __init__(
        self,
        topic_filter,
        callback: Callable[[str, Message], None],
        max_concurrency: int = None,
        max_messages: int = None,
    ):
        super().__init__(topic_filter, callback, max_concurrency, max_messages)

    def parse_raw_payload(self, event):
        try:
            message = Message.from_bytes(event.message.payload)
            logger.debug(
                "IoT Core: decoded EdgeCommons protobuf message on topic %s",
                event.message.topic_name,
            )
            return event.message.topic_name, message
        except ValueError as error:
            logger.warning(
                "Problem decoding IoT Core payload into EdgeCommons protobuf Message on topic %s: "
                "%s. Ignoring message.",
                event.message.topic_name,
                error,
            )
            return None

File: libs/java/src/main/java/com/mbreissi/edgecommons/messaging/providers/greengrass/IotCoreSubscriptionHandler.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.mbreissi.edgecommons.messaging.providers.greengrass;


import com.mbreissi.edgecommons.messaging.Message;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import oshi.util.tuples.Pair;
import software.amazon.awssdk.aws.greengrass.model.IoTCoreMessage;
import software.amazon.awssdk.aws.greengrass.model.MQTTMessage;

import java.util.function.BiConsumer;

public class IotCoreSubscriptionHandler extends SubscriptionHandler<IoTCoreMessage>
{
    protected static final Logger LOGGER = LogManager.getLogger(IotCoreSubscriptionHandler.class);

    public IotCoreSubscriptionHandler(String topicFilter, BiConsumer<String, Message> callback, int maxConcurrency, int maxMessages)
    {
        super(topicFilter, callback, maxConcurrency, maxMessages);
    }

    @Override
    Pair<String, Message> parseRawPayload(IoTCoreMessage iotCoreMessage)
    {
        LOGGER.debug("IoT Core message received on subscription to topic filter '{}'", topicFilter);
        Pair<String, Message> retVal = null;
        try
        {
            String topic = iotCoreMessage.getMessage().getTopicName();
            MQTTMessage mqttMessage = iotCoreMessage.getMessage();
            Message msg = Message.fromBytes(mqttMessage.getPayload());
            retVal = new Pair<>(topic, msg);
        } catch (IllegalArgumentException e) {
            LOGGER.warn("Problem decoding IoT Core payload into EdgeCommons protobuf Message on topic {}: {}.  Ignoring message.",
                    topicFilter, e.toString());
        } catch (Exception e) {
            LOGGER.error("Unexpected error while parsing IoT Core payload: {}. Ignoring message", e.toString());
        }
        return retVal;
    }
}
